### ISSUE TITLE
sys.net.Host not supports IPv6

### ISSUE THREAD

**Issue by Deltachaos** [HaxeFoundation/haxe#3122]
The Host class API is not able to handle IPv6 adresses. I currently dont have any idea how a better API should look like. May it would be a good idea to deprecate the Host class and replace it with an better interface. Any suggestions?

**Comment by ProPuke** [comment]
IPv6 is now becoming a requirement for iOS apps
http://www.internetsociety.org/deploy360/blog/2015/06/apple-will-require-ipv6-support-for-all-ios-9-apps/

**Comment by Annwenn** [comment]
http://www.internetsociety.org/deploy360/blog/2015/06/apple-will-require-ipv6-support-for-all-ios-9-apps/
Now would be a good time to re-open it ;)

**Comment by mcheshkov** [comment]
How about make Host opaque, or hide ip field at least, and/or make it haxe.io.Bytes.
stdlib API uses Host itself, and only cpp and neko sys.net.Socket uses ip field.
In native code neko and cpp are similar, both can accept haxe.io.Bytes at Haxe side (as Socket.readBytes) and use it to pass/store whole sockaddr.sa_data.

**Comment by eminfedar** [comment]
is this issue solved? I think IPv6 is important topic.

**Comment by Mikaela** [comment]
> is this issue solved?

I believe this issue would be closed if it was fixed.

> I think IPv6 is important topic.

I hope everyone thought that way, but most of people are ignoring it entirely, "lets just add CGN after CGN after CGN .. .after CGN." I am very happy that Apple made the policy of requiring IPv6 support from all App Store apps and also improving their Happy Eyeballs.

By the way, isn't iOS9 which requires apps to have IPv6 support released the day after tomorrow, 15th?

**Comment by realcraig** [comment]
As of June 1, 2016 Apple has started enforcing IPv6 support in iOS app submissions. Does anyone know the current status of IPv6 address support in Haxe?

**Comment by ncannasse** [comment]
Is someone interested in contributing it ? This would require some changes but should be feasible without much changes in the API. The only breaking change will be to change our ip representation from Int32 to Bytes so it can hold either 4 bytes or 16 bytes of IPV6. Also, i'm not familiar with the IPv6 C API but from the following article it seems pretty similar to IPv4

http://long.ccaba.upc.edu/long/045Guidelines/eva/ipv6.html
http://tldp.org/HOWTO/Linux+IPv6-HOWTO/chapter-section-using-api.html

**Comment by ngrebenshikov-mhe** [comment]
Now it's unable to pass the review step on the AppStore due to the IPv6 issue. Is there any update here? Any estimations about delivery dates?

**Comment by ngrebenshikov** [comment]
Now it's unable to pass the review step on the AppStore due to the IPv6 issue. Is there any update here? Any estimations about delivery dates?

**Comment by ncannasse** [comment]
Ping @hughsando

**Comment by hughsando** [comment]
Can I change the type of the 'ip' variable without breaking anything?
Maybe I could add a private ipv6 variable and use that internally and not break anything, but still resolve correctly.

**Comment by thamthang** [comment]
Is there any update here please?

**Comment by ncannasse** [comment]
@hughsando yes I think we need some kind of dual usage of the class so it uses either ipv4 or v6, we could have an extra public ipv6 : Bytes field in Host for that

**Comment by hpgood** [comment]
@ncannasse  I had connect server with ipv6,by modify the three file, with your suggestion.
modify the three files is:
\haxe\std\cpp\_std\sys\net\Host.hx
\haxe\std\cpp\_std\sys\net\Socket.hx
\haxe\lib\hxcpp\3,3,49,2\project\libs\std\Socket.cpp

**Comment by hpgood** [comment]
the follow files is modified by my team:
[Socket.cpp.txt](https://github.com/HaxeFoundation/haxe/files/550246/Socket.cpp.txt)
[Host.hx.txt](https://github.com/HaxeFoundation/haxe/files/550250/Host.hx.txt)
[Socket.hx.txt](https://github.com/HaxeFoundation/haxe/files/550249/Socket.hx.txt)

**Comment by andyli** [comment]
@hpgood Thanks for looking into that. Would you submit a PR instead? It is hard to review currently.

**Comment by thamthang** [comment]
@hpgood I've tried your 3 new files but got the errors below while trying to build ios legacy
`./src/haxe/Http.cpp:267:5: error: use of undeclared identifier 'sys'
                                sys::net::Socket tmp8 = sys::net::Socket_obj::__new();          HX_STACK_VAR(tmp8,"tmp8");
                                ^
./src/haxe/Http.cpp:267:29: error: use of undeclared identifier 'sys'
                                sys::net::Socket tmp8 = sys::net::Socket_obj::__new();          HX_STACK_VAR(tmp8,"tmp8");
                                                        ^
./src/haxe/Http.cpp:904:5: error: use of undeclared identifier 'sys'
                                sys::net::Host tmp37 = sys::net::Host_obj::__new(tmp36);                HX_STACK_VAR(tmp37,"tmp37");
                                ^
./src/haxe/Http.cpp:904:28: error: use of undeclared identifier 'sys'
                                sys::net::Host tmp37 = sys::net::Host_obj::__new(tmp36);                HX_STACK_VAR(tmp37,"tmp37");
                                                       ^
./src/haxe/Http.cpp:914:5: error: use of undeclared identifier 'sys'
                                sys::net::Host tmp35 = sys::net::Host_obj::__new(host);         HX_STACK_VAR(tmp35,"tmp35");
                                ^
./src/haxe/Http.cpp:914:28: error: use of undeclared identifier 'sys'
                                sys::net::Host tmp35 = sys::net::Host_obj::__new(host);         HX_STACK_VAR(tmp35,"tmp35");`

**Comment by hpgood** [comment]
@thamthang
after modify the 3 new files,
goto the hxcpp  directory
cd project
neko build.n ios

rebuild the hxcpp project

**Comment by thamthang** [comment]
I've tried these 2 commands but still got the error
neko build.n ios
lime rebuild hxcpp

**Comment by hpgood** [comment]
May be ,you using the version of hxcpp(3,3,49,2)  and haxe (3,2,1)  is diference from me.
But i think the code is similar to modify.  could give some more detail information:
haxelib list
compile cmd  etc.

**Comment by aW4KeNiNG** [comment]
A pull request would be nice =)

I have the same problem here. Apple rejects my app because of that :(

**Comment by ncannasse** [comment]
@hughsando this looks like a serious issue

**Comment by picasso28** [comment]
Hi Guys,

I have just faced this same problem. My company has developed an App for the National Health Department here and Apple has rejected the App  :( :( The deadline to deliver the App is the end of this year... What can we do? Any chance that you guys can deliver a solution before that date? we have invested a lot of time and money in the App and we are now in big troubles :( :( Thank you for your support!!

**Comment by ngrebenshikov** [comment]
@picasso28, we've overwritten haxe.HTTP.request by using iOS system features - https://github.com/jiveui/openfl-ios-networking

Notice: it only works with the OpenFL in legacy mode. May be the idea will help you.

**Comment by picasso28** [comment]
@ngrebenshikov , thank you mate! We will give it an try and let you know how it goes ;) Thanks again

**Comment by PaulGene** [comment]
The new iOS requirement makes this super important to fix.

**Comment by hughsando** [comment]
Well, this is kindof done, but it is not really very good within the existing "core" api.
hxcpp will resolve an address with ipv6 if it can't with ip v4, and then connect if it can.

But if we are going to follow the "native" api, then you need to specify the socket type at creation time, ie "new Ipv6Socket".  But ideally you would need to do this based on whether the host is a v4 of v6 connection, which is unknown by the core api.

You may also want to specify whether you want the Host as a ipv4 or ipv6 host - eg, if you are a server.  There is currently no api for this.

I think the api should be something like: new Host( name, port, mode ) when mode is "ipv4" "ipv6" or "don't care".

Then the "new Socket + connect" or "new Socket + bind" should be dumped, and it should be
var socket = myHost.connect() or myHost.bind().  Leaving the name resolution until the connection may be better too, since the examples show how to iterate over possible resolutions until a connection can be made.  Possibly add a "myHost.getIpv4" and "myHost.getIpv6".

There are also types other than tcpip and udp (eg, bluetooth) that may be interesting.

So it is hacked as best I can do from within 1 target, but the api needs a work-over with input from the other targets.

**Comment by ncannasse** [comment]
@hughsando my proposal would be:

```haxe
class Host {

     static var DEFAULT_IPV6 = false; // can be set to true by user to always default to IPV6
 
     var ip(default,null) : Int; // v4 (or 0)
     var ipv6(default,null) : haxe.io.Bytes; // v6 (or null)
     function new( name : String, ?ipv6 : Bool ) : Void;

}
```

This way you can globally turn an IPV4 app into using IPV6 by switching the static, since you cannot easily modify the way sockets are created in Http for instance.

And for socket connect/bind, I propose we handle it at Haxe level by delaying the underlying socket creation, without modifying the API.

**Comment by hughsando** [comment]
Yes, I think delaying the creation is best - I wonder why it was not done
like this in the first place.
Could it be to do with the input/output streams being needed before the
connect? Or maybe on a "fork" closing one end and using the other (or am I
thinking of 'pipe' here).  So this would change, although not sure if
anyone is relying on this.

In the v6 api there is room for a port when resolving the name - so this
would imply that resolving a host followed by connecting to a port is not
the most accurate thing to do.  There is also room for a 'service' - not
exactly sure what this is, maybe "http" etc.  This may need to be accounted
for too.  I guess there might be these optional args on the Host class
constructor if we are going to change to api.  I'm not sure how much this
is used in practice.

I don't think there is a need for the static "default ipv6" for the http
case.  The hxcpp code will try to resolve ipv4 and if it does, it should
work.  And if it does not, then it tries ipv6 an if this works, then it can
use it.  So http should work no matter what.  The main use for the default
is if you are creating a server on your own machine - do you use a v6 of v4
address?  To specify this, I think a default arg is better since you might
want to do both and a one-off static that you need to change between calls
is just going to cause problems. So the "?ipv6 : Bool" is enough. Also, I
think "localhost" needs something like this.

Since the 'core' classes can be hard to change, I suggest not using default
for properties, ie "var ip(get,null)". This allows the host class to
implement it how it wants instead of forcing the implementation to resolve
at constructor time.

**Comment by aW4KeNiNG** [comment]
Can someone confirm that it works? I've tested it and it doesn't work yet. I'm using a Mac from LAN with Shared Internet and NAT64 network.

Haxe 3.4 final release + hxcpp git.

Thanks in advance.

**Comment by aW4KeNiNG** [comment]
Any help? 

Currently my business is stuck without an iOS version. My clients are pushing me for a iOS version because the Android version is out. And maybe I lost them. I don't want to give pressure, but Hx can't be productive/competitive if we can't launch an iOS application with Internet connection. 

I really wish to help you with a fix from me, but i think you could fix it in a quick way because i haven't worked with haxe core before.

For me, it isn't working with IPv6 connected to MacOS with Sharing Internet/NAT64. I can't test in another situation. So, can someone check if it works?

Thanks in advance.

**Comment by hpgood** [comment]
my app in appstore can connect ipv6 network with my fix.(build in legacy mode)
I will test with new haxe version tomorrow ,and tell you the result.

**Comment by aW4KeNiNG** [comment]
Thanks @hpgood . Currently I'm using OpenFL 4, so legacy mode isn't useful for me.

**Comment by hpgood** [comment]
@aW4KeNiNG  ,I know,so I try to update it in next mode.

**Comment by hughsando** [comment]
There is code in there, so I will need more information about how it is that it does not work.
First test if your LAN/device combo supports it, for example try http://ipv6-test.com/, or ```ping -6 ipv6.test-ipv6.com```
If this does not work, then I have no hope of making it work.
Then try something like:
```haxe
   trace( new sys.net.Host("ipv6.test-ipv6.com") );
```
And see what you get

**Comment by hpgood** [comment]
@aW4KeNiNG  @hughsando

I found that ,the haxe 3.4 api (with hxcpp 3.4.49 ,openfl 4.8.1,lime 3.7.4,in next mode) is connect fine with ipv6 network!!
my build command is:
openfl test ios -Dnext -x -v

There are some ipv6 implementations in "NativeSocket.hx"(https://github.com/HaxeFoundation/haxe/blob/development/std/cpp/NativeSocket.hx)

so, this issues should close.

        var ipv6 = NativeSocket.host_resolve_ipv6("www.baidu.com");
	if(ipv6!=null){
		var ipv6_str = ipv6.join(":");
		Log.trace("ipv6_str->" + ipv6_str );
	}else{
		Log.trace(" ipv6 is null"   );
	}
attention：
  cannot using ipv4 address in ipv6 wifi network.
for example:   
      var urlrq:URLRequest = new URLRequest("http://192.168.1.10/1.txt");  //it is wrong in ipv6 network.
should replace by:
      var urlrq:URLRequest = new URLRequest("http://some_other_domain/1.txt");  //it is right.
      or
      var urlrq:URLRequest = new URLRequest("http://1030::C9B4:FF12:48AA:1A2B/1.txt");  //it is right.

**Comment by aW4KeNiNG** [comment]
Thanks! Good to know.

I will try later again. Anyway, I will try to be lucky with Apple.

**Comment by aW4KeNiNG** [comment]
Hi!

The application have been rejected 3 times in 2 weeks (2 with TestFlight and 1 with Final Distribution). It isn't working for me, and I need help @hughsando .

The Apple rejection message:

```
Performance - 2.1


We discovered one or more bugs in your app when reviewed on iPad and iPhone running iOS 10.2.1 on Wi-Fi connected to an IPv6 network.

Specifically, the app displayed an error at login.

Please see attached screenshots for details.

Next Steps

Please run your app on a device while connected to an IPv6 network (all apps must support IPv6) to identify any issues, then revise and resubmit your app for review.
```
<img width="300" alt="attachment-3606294957825341102screenshot_0316_110327" src="https://cloud.githubusercontent.com/assets/1063719/24011951/91c374bc-0a7c-11e7-9497-6a182919a864.png">

I can replicate the same problem with MacOS in Sharing Internet/NAT64 option and iOS/Windows/Android app (but it isn't a true IPv6 in Internet, it is only for local ip).

IPConfig:

```
Wireless LAN adapter Wi-Fi 7:

   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Intel(R) Centrino(R) Wireless-N 2230 #6
   Physical Address. . . . . . . . . : 68-5D-43-12-0A-70
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes
   IPv6 Address. . . . . . . . . . . : 2001:2:0:aab1:c059:121a:614b:1d2a(Preferred)
   Temporary IPv6 Address. . . . . . : 2001:2:0:aab1:243f:9627:73ff:a3d4(Preferred)
   Link-local IPv6 Address . . . . . : fe80::c059:121a:614b:1d2a%50(Preferred)
   Autoconfiguration IPv4 Address. . : 169.254.29.42(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.0.0
   Default Gateway . . . . . . . . . : fe80::f45c:89ff:fe49:dc64%50
   DHCPv6 IAID . . . . . . . . . . . : 728259907
   DHCPv6 Client DUID. . . . . . . . : 00-01-00-01-1C-FA-C0-50-68-5D-43-12-0A-70
   DNS Servers . . . . . . . . . . . : 2001:2:0:aab1::1
   NetBIOS over Tcpip. . . . . . . . : Enabled

```

`http://ipv6-test.com/` result:

![capture1](https://cloud.githubusercontent.com/assets/1063719/24012124/2af6c0da-0a7d-11e7-9f9b-0dff27ba4cfe.JPG)


The next simple app is reporting me an error:

```
trace( new sys.net.Host("ipv6.test-ipv6.com") );
```

![capture](https://cloud.githubusercontent.com/assets/1063719/24012045/e578f42e-0a7c-11e7-9ed6-f192a1450e89.JPG)


My config:

```
C:\Users\Pablo>haxelib list
hxcpp: 3.2.205 3.3.47 3.3.49 3.4.2 [3.4.49]
lime: 2.9.0 2.9.1 3.3.0 3.4.1 3.5.0 3.5.1 3.5.2 3.6.0 3.6.2 3.7.0 3.7.2 3.7.3 [3.7.4]
openfl: 3.4.0 3.5.3 3.6.0 3.6.1 4.3.0 4.3.1 4.4.1 4.5.0 4.5.1 4.5.2 4.5.3 4.6.0 4.7.0 4.7.3 4.8.0 [4.8.1]

C:\Users\Pablo>haxe -version
3.4.0
```

For iOS build I'm using MacOS Sierra 10.12 + XCode 8.2.1. But the problem occurs in Windows and Android too. 


Thanks in advance.

**Comment by hughsando** [comment]
from the ipv6-test results, it looks like you browser does not support ipv6 on your network, so hxcpp does not really have a chance of working either.
I guess we need a situation where the ipv6 test works, and hxcpp does  not, and we can then work out where they differ.
I do not really have the expertise to look into this.
The best solution might be to use the native ios network api, since this must satisfy apple.

**Comment by aW4KeNiNG** [comment]
Thanks @hughsando .
Maybe it couldn't work in my PC, but it should work in iOS or Android. Apple is testing with IPv6, iOS devices and the last Haxe (3.4.0)/HxCPP (3.4.49) version (some days ago a new version of Haxe/HxCPP is released, but it doesn't fix/improve IPv6). The problem isn't my network/configuration because Apple has the same problem too and Apple can't validate any app if it doesn't work in its system.

@hpgood , how is it working for you? Are you using your custom patch? Are you submitted an application with Internet required into Apple Store?

**Comment by hpgood** [comment]
@aW4KeNiNG 
I have a patch maybe help you:
https://github.com/HaxeFoundation/haxe/blob/development/std/cpp/_std/sys/net/Host.hx

public function new( name : String ) : Void {
		host = name;
      try {
			ip = NativeSocket.host_resolve(name);
      }
      catch(e:Dynamic)
      {
		   ipv6 = NativeSocket.host_resolve_ipv6(name);
      }
	}

change to:


public function new( name : String ) : Void {
		host = name;
      try { 
             ipv6 = NativeSocket.host_resolve_ipv6(name);
      }catch(e:Dynamic){
		 ipv6=null;  
      }
      try {
	    if(ipv6==null) {
                ip = NativeSocket.host_resolve(name);
            }
      } catch(e:Dynamic) {   }
   }

my app now is testing with testflight ,it will submit to app store in the next day. and my app only working with Internet .

**Comment by aW4KeNiNG** [comment]
Thanks @hpgood !. If it is accepted in TestFlight, could you upload a Pull Request with the change? It means that it should works into App Store too, because my app isn't accepted in TestFlight or App Store. 

Also let us know about your state in TestFlight/App Store.

**Comment by aW4KeNiNG** [comment]
@hpgood any news? Has the App been accepted in TestFlight?

**Comment by hpgood** [comment]
yes,it allway accept in testflight ,don't worry,i will update the news here.

**Comment by aW4KeNiNG** [comment]
Thank you very much!! :)

I will wait to the pull request to test myself. @hughsando can you take a look?

**Comment by aW4KeNiNG** [comment]
@hpgood, @hughsando , I have tested your code and now it works with IPv6 with Mac NAT64 wifi, but it didn't work with IPv4.

Actually, with IPv4, it gets a valid value in the `ipv6` variable, so it doesn't use the ipv4 workflow (with `ip` variable).

**Comment by hpgood** [comment]
what is your url address? you should use domain replace the ip (number) address.

**Comment by aW4KeNiNG** [comment]
For example, `http://www.google.com` didn't work.

**Comment by hughsando** [comment]
Looking a the case where you are trying for a ipv6 address.
The code first tries the ipv4 address.  On iOS(mac?), it looks like this get some kind of valid ipv4 address (no error thrown), but then this does not subsequently work with the socket call.  I'm wondering what this address is.  Is it some tunnel/relay address which does not support https, or maybe some fake address like "0.0.0.0" which needs to be interpreted differently.  Or maybe an alternate domain?
I also see that the ipv6 code returns null, rather than throwing - which complicates your patch. I should fix this.
If you try ipv6 first, it looks like the DNS works (return valid socket) but the socket code can't use it if the OS is not set up to use it.
I wonder how it would go if we did not even try ipv4 on ios, or always try ipv6 first based on operating system.  This could be done using a conditional compile in hxcpp to reject ipv4 connections on ios if ipv6 addresses work.

**Comment by hughsando** [comment]
Looking at ios - it might be possible to never return a ipv4 address (from ios 4.1+), since I think it should make a relay address if it has to.  This would simplify testing since we would not need to worry about which one to try first.  Would need to test things like "localhost" and "192.1687.1.1" etc, but if it works, it is a simple solution.

**Comment by aW4KeNiNG** [comment]
@hughsando, yesterday I worked in a fix, and it is working now, with IPv4 and IPv6. Today it was accepted in TestFlight.

Let me create a PR.

**Comment by aW4KeNiNG** [comment]
I have created the PR #6160. Maybe it is a temporal solution, but it is working now. My app is already accepted in TestFlight, and it works with IPv6 and IPv4 (the @hpgood solution didn't work with IPv4 for me).

Currently, it checks to connect with IPv4, if it receives an EOF error, then it checks with IPv6 (if available). 

Anyway, thanks for your support @hughsando and @hpgood.

**Comment by aW4KeNiNG** [comment]
My application is already accepted in App Store with the changes :)

**Comment by starburst997** [comment]
@aW4KeNiNG  Thanks for PR #6160 ! All seems to works, however Https seems to failed, Http is OK. Are you able to use https or is it just me?

**Comment by starburst997** [comment]
Maybe it's my network configuration or something but with PR #6160, request on IPv6 seems fast on all target except iOS where it can takes a minute each time. (it's works tho!)

**Comment by starburst997** [comment]
About the slow request on iOS, just need to change the order as in PR #6142

**Comment by starburst997** [comment]
Actually, PR #6142 seems to do the job just fine!

**Comment by aW4KeNiNG** [comment]
Hi @starburst997 .

I don't have any problems with iOS and IPv6. It is fast like IPv4. Check the PR #6142 because it didn't work with IPv4 anymore. It is valid to pass the Apple Store, but if you check the connection with IPv4 it didn't work for me.

**Comment by starburst997** [comment]
Weird it did for me, but maybe it was a cache issue or something, I'm going to do some more tests.

**Comment by starburst997** [comment]
I created a repo to do some tests on IPV6 and SSL.

https://github.com/starburst997/haxe-openfl-ipv6-ssl

I test Http, [Akifox](https://github.com/yupswing/akifox-asynchttp) (because it uses Socket) and URLLoader

I'm using OpenFL 4 (didn't try other version)

To test IPV6 I used this guide: [Mac Internet Sharing](https://developer.apple.com/library/content/documentation/NetworkingInternetWeb/Conceptual/NetworkingOverview/UnderstandingandPreparingfortheIPv6Transition/UnderstandingandPreparingfortheIPv6Transition.html#//apple_ref/doc/uid/TP40010220-CH213-SW16)

***My results are as follow:***

- [IPV6 No PR](https://github.com/starburst997/haxe-openfl-ipv6-ssl/tree/master/results/IPV6_NOPR.md)
- [IPV6 #6142](https://github.com/starburst997/haxe-openfl-ipv6-ssl/tree/master/results/IPV6_PR6142.md)
- [IPV6 #6160](https://github.com/starburst997/haxe-openfl-ipv6-ssl/tree/master/results/IPV6_PR6160.md)
- [IPV4 #PR](https://github.com/starburst997/haxe-openfl-ipv6-ssl/tree/master/results/IPV4_NOPR.md)
- [IPV4 #6142](https://github.com/starburst997/haxe-openfl-ipv6-ssl/tree/master/results/IPV4_PR6142.md)
- [IPV4 #6160](https://github.com/starburst997/haxe-openfl-ipv6-ssl/tree/master/results/IPV4_6160.md)

*Basically*:

HTTPS on IPV6 doesn't works, ***EOF***

Both #6142 and #6160 fixes IPV6 connectivity.

Both #6142 and #6160 gives the same results except that I get ***timeout*** (like 60+ sec between each requests) with #6160 (only on IOS, I have an iPhone 4s) which I don't get with #6142.

On IPV4 all seems fine, except for some reason, OpenFL's URLLoader doesn't like my Server's SSL (Error: 35), but that's another issue...

IPV4 works fine using both PR

***On IOS***: OpenFL's URLLoader seems to be able to do the job just fine, IPV6 + SSL, and it's the only one working (but failed with Error: 6 on other target)

**Comment by hughsando** [comment]
So - #6142 seems good on ios.  If I do ipv6 first on ios, and second on all
others does that seem safest?
